jena-shacl/src/main/java/org/apache/jena/shacl/validation/ReportEntry.java

```diff
@@ -0,0 +1,303 @@
+
+    public ReportEntry sourceConstraintComponent(Node component) {
+        this.sourceConstraintComponent = component;
+        return this;
+    }
+
+    public Node detail() { return detail; }
+
+    public ReportEntry detail(Node n) {
+        detail = n;
+        return this;
+    }
+
+    public ReportEntry message(Node node) {
+        if ( node != null )
+            messages.add(node);
+        return this;
+    }
+
+    public ReportEntry message(String msg) {
+        if ( msg != null )
+            messages.add(NodeFactory.createLiteral(msg));
+        return this;
+    }
+
+    public String message() {
+        if ( messages() != null && messages().size() == 1 )
+            return asString(CollectionUtils.oneElt(messages));
+        else {
+            StringJoiner sj = new StringJoiner(",", "[","]");
+            for ( Node n : messages )
+                sj.add(NodeFmtLib.displayStr(n));
+            return sj.toString();
+        }
+    }
+
+    public Collection<Node> messages() { return messages; }
+
+    private static String asString(Node message) {
+        if ( message == null )
+            return null;
+        if ( message.isLiteral() )
+            return message.getLiteralLexicalForm();
+        else if ( message.isURI() )
+            return message.getURI();
+        else
+            throw new ShaclException("Not a literal or URI: "+message);
+    }
+
+    public Severity severity() { return severity; }
+
+    public ReportEntry severity(Severity severity) {
+        this.severity = severity;
+        return this;
+    }
+
+    public ReportEntry severity(Node severity) {
+        this.severity = (severity == null) ? null : Severity.create(severity);
+        return this;
+    }
+
+    public void generate(Model model, Resource report) {
+        Resource entry = model.createResource();
+        entry.addProperty(RDF.type, SHACLM.ValidationResult);
+        report.addProperty(SHACLM.result, entry);
+
+        Function<Node, RDFNode> rdfNode = (gn)->ModelUtils.convertGraphNodeToRDFNode(gn, entry.getModel());
+        Function<String, RDFNode> str = (s)->entry.getModel().createLiteral(s);
+
+        // Section 3.6.2
+        // The properties sh:focusNode, sh:resultSeverity and
+        // sh:sourceConstraintComponent are the only properties that are mandatory
+        // for all validation results.
+
+        Objects.requireNonNull(focusNode,"focus node");
+        //Objects.requireNonNull(message,"message");
+        Objects.requireNonNull(severity,"severity");
+        Objects.requireNonNull(sourceConstraintComponent, "sourceConstraintComponent");
+
+        entry.addProperty(SHACLM.focusNode,       rdfNode.apply(focusNode));
+        entry.addProperty(SHACLM.resultSeverity,  rdfNode.apply(severity.level()));
+        entry.addProperty(SHACLM.sourceConstraintComponent, rdfNode.apply(sourceConstraintComponent));
+
+        if ( messages != null ) {
```

Comment:
It's never `null` as it's a private/final `ArrayList` initialized only in this class (other methods use `messages` without defensive check)